src/writer.rs

```diff
@@ -149,51 +212,61 @@ where
 /// Write the data in 'f' order.
 ///
 /// Like NiBabel, we iterate by "slice" to improve speed and use less memory.
-fn write_data<A, S, D, W>(writer: &mut W, header: NiftiHeader, data: &ArrayBase<S, D>) -> Result<()>
+fn write_data<T, D, W>(writer: &mut W, header: NiftiHeader, data: ArrayView<T, D>) -> Result<()>
 where
-    S: Data<Elem = A>,
-    A: Copy,
-    A: DataElement,
-    A: Div<Output = A>,
-    A: FromPrimitive,
-    A: PodTransmutable,
-    A: ScalarOperand,
-    A: Sub<Output = A>,
+    T: Clone + PodTransmutable,
+    T: Div<Output = T>,
+    T: FromPrimitive,
+    T: PodTransmutable,
+    T: ScalarOperand,
+    T: Sub<Output = T>,
     D: Dimension + RemoveAxis,
     W: WriteBytesExt,
 {
-    // Need the transpose for fortran ordering used in nifti file format.
-    let data = data.t();
-
-    let mut write_slice = |data: Array<A, D::Smaller>| -> Result<()> {
-        let len = data.len();
-        let arr_data = data.into_shape(len).unwrap();
-        let slice = arr_data.as_slice().unwrap();
-        writer.write_all(guarded_transmute_to_bytes_pod_many(slice))?;
-        Ok(())
-    };
-
     // `1.0x + 0.0` would give the same results, but we avoid a lot of divisions
     let slope = if header.scl_slope == 0.0 {
         1.0
     } else {
         header.scl_slope
     };
     if slope != 1.0 || header.scl_inter != 0.0 {
-        let slope = A::from_f32(slope).unwrap();
-        let inter = A::from_f32(header.scl_inter).unwrap();
+        let slope = T::from_f32(slope).unwrap();
+        let inter = T::from_f32(header.scl_inter).unwrap();
```

Comment:
I think this part of the function should be reiterated (either now or in a separate PR), since it does not exactly resemble the current logic employed when reading transformed data. The [`data::element`](https://github.com/Enet4/nifti-rs/blob/master/src/volume/element.rs#L20) module provides multiple linear transformation strategies for each valid element data type to prevent precision loss. In this particular code, an `scl_slope` of `0.5` (haven't even seen real data with this, but it's a valid one nevertheless) would turn all voxel values to 0 if we pass an ndarray of integers. We might need a bit more code than just reusing this module though, since the transformation strategies currently live in immaterial types.

Reply:
Ouch. True. I'll check if I can copy your design.